File: agentdock-plugin/src/test/kotlin/com/agentdock/terminal/CommandRendererTest.kt
package com.agentdock.terminal

import com.agentdock.model.AgentSession
import com.agentdock.model.CLIProvider
import com.agentdock.model.ProviderCommandContext
import com.agentdock.util.OperatingSystem
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs

class CommandRendererTest {
    private val provider = CLIProvider.defaultProviders().first { it.id == CLIProvider.CODEX_ID }

    @Test
    fun `renders provider session id`() {
        val session = AgentSession(
            name = "Fix tests",
            providerId = provider.id,
            cwd = "/tmp/project",
            providerSessionId = "codex_123"
        )

        val result = CommandRenderer().render(
            "codex resume {{providerSessionId}}",
            context(session)
        )

        assertIs<CommandRenderResult.Success>(result)
        assertEquals("codex resume codex_123", result.command)
    }

    @Test
    fun `renders executable variable`() {
        val session = AgentSession(name = "Fix tests", providerId = provider.id, cwd = "/tmp/project")

        val result = CommandRenderer().render(
            "{{executable}} --version",
            context(session)
        )

        assertIs<CommandRenderResult.Success>(result)
        assertEquals("codex --version", result.command)
    }

    @Test
    fun `removes optional missing provider session id`() {
        val session = AgentSession(name = "Fix tests", providerId = provider.id, cwd = "/tmp/project")

        val result = CommandRenderer().render(
            "{{executable}} resume {{providerSessionId?}}",
            context(session)
        )

        assertIs<CommandRenderResult.Success>(result)
        assertEquals("codex resume", result.command)
    }

    @Test
    fun `fails when provider session id is required but missing`() {
        val session = AgentSession(name = "Fix tests", providerId = provider.id, cwd = "/tmp/project")

        val result = CommandRenderer().render(
            "codex resume {{providerSessionId}}",
            context(session)
        )

        assertIs<CommandRenderResult.MissingVariable>(result)
        assertEquals("providerSessionId", result.variable)
    }

    @Test
    fun `escapes variables with spaces`() {
        val session = AgentSession(
            name = "Fix quoted task",
            providerId = provider.id,
            cwd = "/tmp/project with spaces",
            providerSessionId = "codex_123"
        )

        val result = CommandRenderer().render("cd {{cwd}}", context(session))

        assertIs<CommandRenderResult.Success>(result)
        assertEquals("cd '/tmp/project with spaces'", result.command)
    }

    @Test
    fun `renders provider specific yolo resume commands`() {
        val expectedCommands = mapOf(
            CLIProvider.CODEX_ID to "codex resume --dangerously-bypass-approvals-and-sandbox session-123",
            CLIProvider.CLAUDE_CODE_ID to "claude --resume session-123 --ide --dangerously-skip-permissions",
            CLIProvider.GEMINI_ID to "gemini --resume session-123 --yolo"
        )

        CLIProvider.defaultProviders().forEach { currentProvider ->
            val session = AgentSession(
                name = "YOLO test",
                providerId = currentProvider.id,
                cwd = "/tmp/project",
                providerSessionId = "session-123"
            )
            val result = CommandRenderer().render(
                currentProvider.yoloResumeCommandTemplate,
                context(session, currentProvider)
            )

            assertIs<CommandRenderResult.Success>(result)
            assertEquals(expectedCommands.getValue(currentProvider.id), result.command)
        }
    }

    private fun context(session: AgentSession, currentProvider: CLIProvider = provider): ProviderCommandContext {
        return ProviderCommandContext(
            provider = currentProvider,
            session = session,
            projectPath = "/tmp/project",
            shell = "/bin/zsh",
            os = OperatingSystem.Mac
        )
    }
}

File: AgentDock-VSCode/test/commandRenderer.test.ts
import { describe, expect, it } from "vitest";
import { renderCommand } from "../src/core/commandRenderer.js";
import { providerById } from "../src/core/providers.js";
import type { AgentSession, ProviderCommandContext } from "../src/core/model.js";

const session: AgentSession = {
  id: "codex:session-id",
  projectPath: "/tmp/project with spaces",
  name: "Session  with  intentional spacing",
  providerId: "codex",
  cwd: "/tmp/project with spaces",
  providerSessionId: "session-id",
  historyFilePath: "/tmp/session.jsonl",
  summary: "Summary",
  createdAt: 1,
  updatedAt: 2,
  pinned: false
};

function context(platform: NodeJS.Platform = "darwin"): ProviderCommandContext {
  return {
    provider: providerById("codex"),
    session,
    projectPath: session.projectPath,
    platform
  };
}

describe("renderCommand", () => {
  it("quotes substituted values without collapsing their internal whitespace", () => {
    expect(renderCommand("{{executable}} --name {{sessionName}}", context())).toEqual({
      ok: true,
      command: "codex --name 'Session  with  intentional spacing'"
    });
  });

  it("removes a missing optional value", () => {
    const result = renderCommand("{{executable}} resume {{providerSessionId?}}", {
      ...context(),
      session: { ...session, providerSessionId: "" }
    });
    expect(result).toEqual({ ok: true, command: "codex resume" });
  });

  it("reports a missing required value", () => {
    const result = renderCommand("{{executable}} --cwd {{cwd}}", {
      ...context(),
      session: { ...session, cwd: "" }
    });
    expect(result).toEqual({ ok: false, missingVariable: "cwd" });
  });

  it("uses Windows-compatible quoting", () => {
    expect(renderCommand("{{executable}} --cwd {{cwd}}", context("win32"))).toEqual({
      ok: true,
      command: 'codex --cwd "/tmp/project with spaces"'
    });
  });
});
